cfme/markers/env_markers/appliance.py

```diff
@@ -0,0 +1,112 @@
+
+    Note:
+        Using the default 'function' scope, each test will be run individually for each appliance
+        before moving on to the next test. To group all tests related to single appliance together,
+        parametrize tests in the 'module' scope.
+
+    Note:
+        testgen for providers now requires the usage of test_flags for collection to work.
+        Please visit http://cfme-tests.readthedocs.org/guides/documenting.html#documenting-tests
+        for more details.
+    """
+    argnames = []
+    argvalues = []
+    idlist = []
+
+    # supported_providers are the ones "supported" in the supportability.yaml file. It will
+    # be a list of DataProvider objects and will be filtered based upon what the test has asked for
+    holder = metafunc.config.pluginmanager.get_plugin('appliance-holder')
+    current_appliance = holder.held_appliance
+    argnames.append(fixture_name)
+    for app_type in app_types:
+        idlist.append(app_type)
+        if app_type == current_appliance.type:
+            argvalues.append([current_appliance])
+        else:
+            msg = (f'No appliance of such type available. needed {app_type}, '
+                   f'exists {current_appliance.type}')
+            argvalues.append(pytest.param(None, marks=pytest.mark.skip(reason=msg)))
+    return argnames, argvalues, idlist
+
+
+class ApplianceEnvironmentMarker(EnvironmentMarker):
+    NAME = 'appliance'
+    DEFAULT = 'regular'
+    CHOICES = ['regular', 'multi-region', 'dummy', 'dev', 'pod', 'upgraded']
+
+    def process_env_mark(self, metafunc):
+
+        # organize by fixture_name kwarg to the marker
+        # iter_markers returns most local mark first, maybe don't need override
+        marks_by_fixture = self.get_closest_kwarg_markers(metafunc.definition)
+        if marks_by_fixture is None:
+            # let's add appliance marker if it's absent
+            marks_by_fixture = {self.DEFAULT: pytest.mark.appliance([self.DEFAULT], scope='module')}
+        # process each mark, defaulting fixture_name
+        for fixture_name, mark in marks_by_fixture.items():
+
+            # mark is either the lowest marker (automatic override), or has custom fixture_name
+            logger.debug(f'Parametrizing appliance env mark {mark}')
+            args = mark.args
+            kwargs = mark.kwargs.copy()
+            scope = kwargs.pop('scope', 'function')
+            indirect = kwargs.pop('indirect', False)
+            filter_unused = kwargs.pop('filter_unused', True)
+            gen_func = kwargs.pop('gen_func', appliances)
+
+            mark_param = args[0] if args else self.DEFAULT
+
+            app_types = []
+            if isinstance(mark_param, (list, tuple)):
+                app_types = mark_param
+            elif mark_param == testgen.ALL:
+                app_types = self.CHOICES
+            elif mark_param == testgen.ONE:
+                app_types.append(random.choice(self.CHOICES))
```

Comment:
I'm thinking testgen.ONE here should default the appliance selection to DEFAULT - otherwise this could randomly pick dev/dummy/upgraded, which I don't think we want to be picked randomly.